File: agent/sync-agent.js
/**
 * sync-agent.js — Nexa Sync Agent
 *
 * Lê o arquivo .db gerado pelo Nexa Clinic (SQLite) e sincroniza
 * os dados com o Supabase (Nexa Insight Web).
 *
 * Tabelas locais  →  Tabelas Supabase
 * ─────────────────────────────────────
 * patients        →  pacientes  (futura)
 * users           →  profiles   (somente leitura — não sobrescreve)
 * [financeiro, repasses, medicos vêm de versões futuras do sistema local]
 *
 * O campo external_id em cada tabela do Supabase serve como chave
 * de idempotência: upsert não duplica registros mesmo rodando várias vezes.
 */

import Database from "better-sqlite3";
import fetch from "node-fetch";
import cron from "node-cron";
import { readFileSync, existsSync } from "fs";
import { resolve } from "path";
import { argv } from "process";

// ─── Config ────────────────────────────────────────────────────────────────

const CONFIG_PATH = resolve("./config.json");

if (!existsSync(CONFIG_PATH)) {
  console.error(
    "❌ config.json não encontrado. Execute: node setup.js"
  );
  process.exit(1);
}

const config = JSON.parse(readFileSync(CONFIG_PATH, "utf-8"));
const { supabase_url, supabase_service_key, empresa_id, db_path, intervalo_minutos, log_level } = config;

const RUN_ONCE = argv.includes("--once");

// ─── Logger ────────────────────────────────────────────────────────────────

function log(level, msg, data) {
  const ts = new Date().toLocaleString("pt-BR");
  const prefix = { info: "ℹ️ ", warn: "⚠️ ", error: "❌", ok: "✅" }[level] ?? "  ";
  console.log(`[${ts}] ${prefix} ${msg}`, data ? JSON.stringify(data) : "");
}

// ─── Supabase helpers ───────────────────────────────────────────────────────

const HEADERS = {
  "Content-Type": "application/json",
  "apikey": supabase_service_key,
  "Authorization": `Bearer ${supabase_service_key}`,
  "Prefer": "resolution=merge-duplicates",
};

async function supabaseUpsert(table, rows) {
  if (!rows.length) return { count: 0 };

  const res = await fetch(`${supabase_url}/rest/v1/${table}`, {
    method: "POST",
    headers: { ...HEADERS, "Prefer": "resolution=merge-duplicates,return=representation" },
    body: JSON.stringify(rows),
  });

  if (!res.ok) {
    const err = await res.text();
    throw new Error(`Supabase ${table}: ${res.status} — ${err}`);
  }

  return { count: rows.length };
}

async function supabaseInsertLog(entry) {
  try {
    await fetch(`${supabase_url}/rest/v1/logs_sincronizacao`, {
      method: "POST",
      headers: HEADERS,
      body: JSON.stringify(entry),
    });
  } catch {
    // log failure não bloqueia o processo
  }
}

// ─── Leitura do banco local ─────────────────────────────────────────────────

function abrirBanco() {
  if (!existsSync(db_path)) {
    throw new Error(`Arquivo .db não encontrado: ${db_path}`);
  }
  return new Database(db_path, { readonly: true });
}

// Converte patients locais para o formato da tabela 'pacientes' do Supabase.
// O campo external_id usa prefixo "local_" + id local para evitar colisão.
function mapearPacientes(rows) {
  return rows.map((r) => ({
    empresa_id,
    external_id: `local_${r.id}`,
    nome: r.name ?? "Sem nome",
    cpf: r.cpf ?? null,
    nascimento: r.birth ?? null,
    telefone: r.phone ?? null,
    email: r.email ?? null,
    convenio: r.insurance ?? null,
    status: r.status ?? "ativo",
    created_at: r.created_at ?? new Date().toISOString(),
  }));
}

// ─── Sincronização principal ────────────────────────────────────────────────

async function sincronizar() {
  const inicio = Date.now();
  log("info", "Iniciando sincronização...");

  let db;
  const resultado = {
    empresa_id,
    tipo: "sincronizacao_completa",
    status: "sucesso",
    mensagem: "",
    payload: {},
  };

  try {
    db = abrirBanco();

    // ── 1. Pacientes ──────────────────────────────────────────────────────
    let contagemPacientes = 0;
    try {
      const pacientesLocais = db.prepare("SELECT * FROM patients").all();
      log("info", `Pacientes encontrados localmente: ${pacientesLocais.length}`);

      if (pacientesLocais.length > 0) {
        const mapeados = mapearPacientes(pacientesLocais);
        const { count } = await supabaseUpsert("pacientes", mapeados);
        contagemPacientes = count;
        log("ok", `Pacientes sincronizados: ${count}`);
      }
    } catch (e) {
      // Tabela pode não existir ainda no Supabase — avisa mas não aborta
      log("warn", "Pacientes: " + e.message);
    }

    // ── 2. Financeiro ─────────────────────────────────────────────────────
    // Quando o sistema local tiver tabela de financeiro, adicionar aqui.
    // Por ora, verifica se a tabela existe e registra.
    let contagemFinanceiro = 0;
    try {
      const temFinanceiro = db
        .prepare("SELECT name FROM sqlite_master WHERE type='table' AND name='financeiro'")
        .get();

      if (temFinanceiro) {
        const finRows = db.prepare("SELECT * FROM financeiro").all();
        log("info", `Financeiro local: ${finRows.length} registros`);

        const mapeados = finRows.map((r) => ({
          empresa_id,
          external_id: `local_${r.id}`,
          tipo: r.tipo ?? "receita",
          descricao: r.descricao ?? r.description ?? "Lançamento importado",
          valor: Number(r.valor ?? r.value ?? 0),
          data: r.data ?? r.date ?? new Date().toISOString().slice(0, 10),
          status: r.status ?? "pago",
          origem: "nexa_clinic_local",
        }));

        if (mapeados.length) {
          const { count } = await supabaseUpsert("financeiro", mapeados);
          contagemFinanceiro = count;
          log("ok", `Financeiro sincronizado: ${count}`);
        }
      } else {
        log("info", "Tabela 'financeiro' não encontrada no banco local (normal nesta versão)");
      }
    } catch (e) {
      log("warn", "Financeiro: " + e.message);
    }

    // ── 3. Médicos / Repasses ─────────────────────────────────────────────
    let contagemMedicos = 0;
    try {
      const temMedicos = db
        .prepare("SELECT name FROM sqlite_master WHERE type='table' AND name='medicos'")
        .get();

      if (temMedicos) {
        const medRows = db.prepare("SELECT * FROM medicos WHERE ativo=1 OR ativo IS NULL").all();
        const mapeados = medRows.map((r) => ({
          empresa_id,
          external_id: `local_${r.id}`,
          nome: r.nome ?? r.name,
          especialidade: r.especialidade ?? r.specialty ?? null,
          crm: r.crm ?? null,
          percentual_repasse: Number(r.percentual_repasse ?? r.percentual ?? 50),
          ativo: true,
        }));

        if (mapeados.length) {
          const { count } = await supabaseUpsert("medicos", mapeados);
          contagemMedicos = count;
          log("ok", `Médicos sincronizados: ${count}`);
        }
      }
    } catch (e) {
      log("warn", "Médicos: " + e.message);
    }

    // ── Resultado ─────────────────────────────────────────────────────────
    const duracao = Date.now() - inicio;
    resultado.mensagem = `Sincronização concluída em ${duracao}ms`;
    resultado.payload = {
      pacientes: contagemPacientes,
      financeiro: contagemFinanceiro,
      medicos: contagemMedicos,
      duracao_ms: duracao,
    };

    log("ok", resultado.mensagem, resultado.payload);

  } catch (e) {
    resultado.status = "erro";
    resultado.mensagem = e.message;
    log("error", "Falha na sincronização: " + e.message);
  } finally {
    if (db) db.close();
    await supabaseInsertLog(resultado);
  }

  return resultado;
}

// ─── Inicialização ──────────────────────────────────────────────────────────

async function main() {
  console.log("\n╔══════════════════════════════════════════╗");
  console.log("║       Nexa Sync Agent v1.0               ║");
  console.log("╚══════════════════════════════════════════╝\n");
  log("info", `Banco local: ${db_path}`);
  log("info", `Supabase: ${supabase_url}`);
  log("info", `Empresa ID: ${empresa_id}`);

  if (RUN_ONCE) {
    log("info", "Modo: sincronização única (--once)");
    await sincronizar();
    process.exit(0);
  }

  // Sincroniza imediatamente ao iniciar
  await sincronizar();

  // Agenda sincronizações periódicas
  const expr = `*/${intervalo_minutos} * * * *`;
  log("info", `Agendado: a cada ${intervalo_minutos} minutos (cron: ${expr})`);

  cron.schedule(expr, async () => {
    await sincronizar();
  });

  log("ok", "Agente rodando. Pressione Ctrl+C para encerrar.\n");
}

main().catch((e) => {
  console.error("Erro fatal:", e);
  process.exit(1);
});

File: agent/sync_agent.py
"""
Nexa Sync Agent — Python
Lê o arquivo .db (SQLite) do Nexa Clinic e sincroniza com o Supabase.

Requisitos: Python 3.8+ (sem instalar nada — usa só biblioteca padrão)
Uso:
  python sync_agent.py          → roda continuamente
  python sync_agent.py --once   → sincroniza uma vez e encerra
"""

import sqlite3
import json
import urllib.request
import urllib.error
import time
import sys
import os
from datetime import datetime, timezone

# ── Config ──────────────────────────────────────────────────────────────────

CONFIG_PATH = os.path.join(os.path.dirname(__file__), "config.json")

def carregar_config():
    if not os.path.exists(CONFIG_PATH):
        print("❌ config.json não encontrado. Execute: python setup.py")
        sys.exit(1)
    with open(CONFIG_PATH, "r", encoding="utf-8") as f:
        return json.load(f)

# ── Logger ───────────────────────────────────────────────────────────────────

def log(nivel, msg, dados=None):
    ts = datetime.now().strftime("%H:%M:%S")
    icone = {"info": "ℹ️ ", "ok": "✅", "erro": "❌", "aviso": "⚠️ "}.get(nivel, "  ")
    linha = f"[{ts}] {icone} {msg}"
    if dados:
        linha += f" {json.dumps(dados, ensure_ascii=False)}"
    print(linha)

# ── Supabase ─────────────────────────────────────────────────────────────────

def supabase_upsert(cfg, tabela, linhas):
    if not linhas:
        return 0
    url = f"{cfg['supabase_url']}/rest/v1/{tabela}"
    dados = json.dumps(linhas).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=dados,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "apikey": cfg["supabase_service_key"],
            "Authorization": f"Bearer {cfg['supabase_service_key']}",
            "Prefer": "resolution=merge-duplicates,return=minimal",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            resp.read()
            return len(linhas)
    except urllib.error.HTTPError as e:
        corpo = e.read().decode("utf-8", errors="replace")
        raise Exception(f"HTTP {e.code} em {tabela}: {corpo}")

def supabase_gravar_log(cfg, entrada):
    url = f"{cfg['supabase_url']}/rest/v1/logs_sincronizacao"
    dados = json.dumps(entrada).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=dados,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "apikey": cfg["supabase_service_key"],
            "Authorization": f"Bearer {cfg['supabase_service_key']}",
            "Prefer": "return=minimal",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            resp.read()
    except Exception:
        pass  # falha no log não interrompe o processo

# ── Leitura do banco local ────────────────────────────────────────────────────

def tabela_existe(cursor, nome):
    cursor.execute(
        "SELECT name FROM sqlite_master WHERE type='table' AND name=?", (nome,)
    )
    return cursor.fetchone() is not None

def sincronizar(cfg):
    inicio = time.time()
    log("info", "Iniciando sincronização...")

    empresa_id = cfg["empresa_id"]
    db_path = cfg["db_path"]

    resultado = {
        "empresa_id": empresa_id,
        "tipo": "sincronizacao_completa",
        "status": "sucesso",
        "mensagem": "",
        "payload": {},
    }

    if not os.path.exists(db_path):
        resultado["status"] = "erro"
        resultado["mensagem"] = f"Arquivo .db não encontrado: {db_path}"
        log("erro", resultado["mensagem"])
        supabase_gravar_log(cfg, resultado)
        return

    conn = None
    try:
        conn = sqlite3.connect(db_path)
        conn.row_factory = sqlite3.Row
        cur = conn.cursor()

        contagem = {"pacientes": 0, "financeiro": 0, "medicos": 0}

        # ── Pacientes ────────────────────────────────────────────────────────
        if tabela_existe(cur, "patients"):
            cur.execute("SELECT * FROM patients")
            linhas = cur.fetchall()
            log("info", f"Pacientes locais: {len(linhas)}")
            if linhas:
                mapeados = [
                    {
                        "empresa_id": empresa_id,
                        "external_id": f"local_{r['id']}",
                        "nome": r["name"] or "Sem nome",
                        "telefone": r["phone"] if "phone" in r.keys() else None,
                        "status": "ativo",
                        "created_at": r["created_at"] or datetime.now(timezone.utc).isoformat(),
                    }
                    for r in linhas
                ]
                contagem["pacientes"] = supabase_upsert(cfg, "pacientes", mapeados)
                log("ok", f"Pacientes sincronizados: {contagem['pacientes']}")

        # ── Financeiro ───────────────────────────────────────────────────────
        if tabela_existe(cur, "financeiro"):
            cur.execute("SELECT * FROM financeiro")
            linhas = cur.fetchall()
            if linhas:
                mapeados = [
                    {
                        "empresa_id": empresa_id,
                        "external_id": f"local_{r['id']}",
                        "tipo": r["tipo"] if "tipo" in r.keys() else "receita",
                        "descricao": r["descricao"] if "descricao" in r.keys() else "Importado",
                        "valor": float(r["valor"] if "valor" in r.keys() else 0),
                        "data": r["data"] if "data" in r.keys() else datetime.now().strftime("%Y-%m-%d"),
                        "status": r["status"] if "status" in r.keys() else "pago",
                        "origem": "nexa_clinic_local",
                    }
                    for r in linhas
                ]
                contagem["financeiro"] = supabase_upsert(cfg, "financeiro", mapeados)
                log("ok", f"Financeiro sincronizado: {contagem['financeiro']}")

        # ── Médicos ──────────────────────────────────────────────────────────
        if tabela_existe(cur, "medicos"):
            cur.execute("SELECT * FROM medicos")
            linhas = cur.fetchall()
            if linhas:
                mapeados = [
                    {
                        "empresa_id": empresa_id,
                        "external_id": f"local_{r['id']}",
                        "nome": r["nome"] if "nome" in r.keys() else r["name"],
                        "especialidade": r["especialidade"] if "especialidade" in r.keys() else None,
                        "crm": r["crm"] if "crm" in r.keys() else None,
                        "percentual_repasse": float(r["percentual_repasse"] if "percentual_repasse" in r.keys() else 50),
                        "ativo": True,
                    }
                    for r in linhas
                ]
                contagem["medicos"] = supabase_upsert(cfg, "medicos", mapeados)
                log("ok", f"Médicos sincronizados: {contagem['medicos']}")

        duracao = round(time.time() - inicio, 2)
        resultado["mensagem"] = f"Concluído em {duracao}s"
        resultado["payload"] = {**contagem, "duracao_s": duracao}
        log("ok", resultado["mensagem"], contagem)

    except Exception as e:
        resultado["status"] = "erro"
        resultado["mensagem"] = str(e)
        log("erro", str(e))
    finally:
        if conn:
            conn.close()
        supabase_gravar_log(cfg, resultado)

# ── Main ─────────────────────────────────────────────────────────────────────

def main():
    print("\n╔══════════════════════════════════════════╗")
    print("║    Nexa Sync Agent v2.0 — Python         ║")
    print("╚══════════════════════════════════════════╝\n")

    cfg = carregar_config()
    log("info", f"Banco local: {cfg['db_path']}")
    log("info", f"Supabase: {cfg['supabase_url']}")
    log("info", f"Empresa: {cfg['empresa_id']}")

    once = "--once" in sys.argv

    sincronizar(cfg)

    if once:
        print("\nModo --once concluído.")
        return

    intervalo = int(cfg.get("intervalo_minutos", 30)) * 60
    log("info", f"Próxima sincronização em {cfg.get('intervalo_minutos', 30)} minutos.")
    log("ok", "Agente rodando. Pressione Ctrl+C para encerrar.\n")

    try:
        while True:
            time.sleep(intervalo)
            sincronizar(cfg)
    except KeyboardInterrupt:
        print("\nAgente encerrado.")

if __name__ == "__main__":
    main()
